case_studies/coadds/coadd_decoder.py

```diff
@@ -0,0 +1,299 @@
+import numpy as np
+from typing import Dict, Optional
+import galsim
+import torch
+from torch import Tensor
+from einops import rearrange
+from bliss.catalog import FullCatalog
+from bliss.datasets.background import ConstantBackground
+from bliss.datasets.galsim_galaxies import GalsimBlends
+from bliss.models.galsim_decoder import (
+    SingleGalsimGalaxyPrior,
+    UniformGalsimGalaxiesPrior,
+    FullCatalogDecoder,
+)
+from bliss.catalog import FullCatalog
+from bliss.models.galsim_decoder import SingleGalsimGalaxyDecoder, load_psf_from_file
+from case_studies.coadds.align_single_exposures import align_single_exposures
+
+
+def _sample_n_sources(max_n_sources) -> int:
+    return int(torch.randint(1, max_n_sources + 1, (1,)).int().item())
+
+
+def _uniform(a, b, n_samples=1) -> Tensor:
+    # uses pytorch to return a single float ~ U(a, b)
+    return (a - b) * torch.rand(n_samples) + b
+
+
+def _add_noise_and_background(image: Tensor, background: Tensor) -> Tensor:
+    image_with_background = image + background
+    noise = image_with_background.sqrt() * torch.randn_like(image_with_background)
+    return image_with_background + noise
+
+
+def _linear_coadd(aligned_images, weight):
+    num = torch.sum(torch.mul(weight, aligned_images), dim=0)
+    return num / torch.sum(weight, dim=0)
+
+
+class CoaddUniformGalsimGalaxiesPrior(UniformGalsimGalaxiesPrior):
+    def __init__(
+        self,
+        single_galaxy_prior: SingleGalsimGalaxyPrior,
+        max_n_sources: int,
+        max_shift: float,
+        num_dithers: int,
+    ):
+        super().__init__(
+            single_galaxy_prior,
+            max_n_sources,
+            max_shift,
+        )
+
+    def sample(self, num_dithers: int) -> Dict[str, Tensor]:
+        """Returns a single batch of source parameters."""
+        n_sources = _sample_n_sources(self.max_n_sources)
+
+        params = torch.zeros(self.max_n_sources, self.dim_latents)
+        params[:n_sources, :] = self.single_galaxy_prior.sample(n_sources)
+
+        locs = torch.zeros(self.max_n_sources, 2)
+        locs[:n_sources, 0] = _uniform(-self.max_shift, self.max_shift, n_sources) + 0.5
+        locs[:n_sources, 1] = _uniform(-self.max_shift, self.max_shift, n_sources) + 0.5
+
+        # for now, galaxies only
+        galaxy_bools = torch.zeros(self.max_n_sources, 1)
+        galaxy_bools[:n_sources, :] = 1
+        star_bools = torch.zeros(self.max_n_sources, 1)
+
+        dithers = torch.distributions.uniform.Uniform(-0.5, 0.5).sample([num_dithers * 2])
+        dithers = dithers.reshape(num_dithers, 2)
```

Comment:
You can simplify this a lot with `super()`
```suggestion
        d = super().sample()
        d['dithers] = torch.distributions.uniform.Uniform(-0.5, 0.5).sample([num_dithers,2])
        return d
```